File: optimal-placement-of-buildings-in-a-grid.js
// Time Complexity : O((H*W)PN) where H and W are height and width respectively, P stands for Permute (order matters),
// and N is the given n;
// Space Complexity : O(H*W)
// Did this code successfully run on Leetcode : Not applicable, not on LeetCode
// Any problem you faced while coding this : No

let minDistance;

/*
 * @param {number} w
 * @param {number} h
 * @param {number} n
 * @return {number[][]}
 */ 

const findMinDistance = (H, W, n) => {
    minDistance = Infinity;
    const matrix = new Array(H);
    for (let i = 0; i < matrix.length; i++) {
        matrix[i] = new Array(W).fill(-1);
    }

    backtrack(matrix, 0, 0, n, H, W);
    return minDistance;
};

const backtrack = (matrix, r, c, n, H, W) => {
    // Base case
    if (n == 0) {
        bfs(matrix, H, W);
        return;
    }

    // Logic
    if (c >= W) {
        r++;
        c = 0;
    }
    for (let i = r; i < H; i++) {
        for (let j = c; j < W; j++) {
            // Action
            matrix[i][j] = 0;
            // Recurse
            backtrack(matrix, i, j + 1, n - 1, H, W);
            // backtrack
            matrix[i][j] = -1;
        }
        c = 0;
    }
}

/**
 * @param {number[][]} matrix
 * @return {number[][]}
 */
// BFS
var bfs = function(matrix, H, W) {
    if (matrix == null || matrix.length == 0) return;

    const q = [];
    const visited = new Array(H);
    const dx = [-1, 0, 1, 0];
    const dy = [0, 1, 0, -1];
    for (let i = 0; i < visited.length; i++) {
        visited[i] = new Array(W).fill(false);
    }

    for (let i = 0; i < H; i++) {
        for (let j = 0; j < W; j++) {
            if (matrix[i][j] == 0) {
                visited[i][j] = true;
                q.push([i, j]);
            }
        }
    }

    let dist = 0;
    
    while (q.length > 0) {
        const countNodeLevel = q.length;

        for (let i = 0; i < countNodeLevel; i++) {
            const curr = q.shift();
            for (let k = 0; k < 4; k++) {
                const x = curr[0] + dx[k];
                const y = curr[1] + dy[k];

                if (isValid(H, W, x, y) && !visited[x][y]) {
                    visited[x][y] = true;
                    q.push([x, y]);
                }
            }
        }
        dist++;
    }

    minDistance = Math.min(minDistance, dist - 1);

};

const isValid = (m, n, x, y) => {
    return x >= 0 && x < m && y >= 0 && y < n;
}

File: OptimalPlacementOfBuildingsInAGrid.java
// "static void main" must be defined in a public class.
public class Main {
    public static void main(String[] args) {
        BuildingPlacement obj = new BuildingPlacement();
        System.out.println(obj.getOptimalPlacementOfBuildings(4,4,1));
    }
}
    class BuildingPlacement {
    int minDistance;
    int[][] dirs;    
    public int getOptimalPlacementOfBuildings(int H, int W, int n)
    {
        //create a grid
        int[][] grid = new int[H][W];
        // initialize all values of the grid to -1
         for (int i = 0; i < H; i++){
             for (int j = 0; j < W; j++){
                 grid[i][j] = -1;
           }    
         }
        // directions array
        dirs = new int[][]{{-1,0},{1,0},{0,-1},{0,1}};
        minDistance = Integer.MAX_VALUE;
        backtrack(grid, 0, 0, H, W, n);
        return minDistance;
    }
    private void backtrack(int[][] grid, int r, int c, int H, int W, int n)
    {
        //base
        if (n == 0){
            bfs(grid, H, W);
            return;
        }
        if (c >= W){
            r++;
            c=0;
        }
         
        //logic
        for ( int i = r; i < H; i++){
            for ( int j = c; j < W; j++){
                     //action
                     grid[i][j] = 0;
                    //recurse
                    backtrack(grid, i, j+1, H, W, n-1);
                    //backtrack
                   grid[i][j] = -1;
               }
            c = 0;
        } 
    }
    // BFS for finding minDistance of each combination
    private void bfs(int[][] grid, int H, int W)
    {
        Queue<int[]> q = new LinkedList();
        boolean[][] visited = new boolean[H][W];
        for(int i = 0; i < H; i++){
            for (int j = 0; j < W; j++){
                if (grid[i][j] == 0){
                    q.add(new int[]{i,j});
                    visited[i][j] = true;
                }
            }
        }
        int distance = 0;
        while (!q.isEmpty()){
            int size = q.size();
            for ( int k = 0; k < size; k++){
              int[] curr = q.poll();
              for (int[] dir : dirs){
                int i = dir[0] + curr[0];
                int j = dir[1] + curr[1];
                if (i<H && i>=0 && j <W && j >=0 && !visited[i][j]){
                    q.add(new int[]{i,j});
                    visited[i][j] = true;
                }           
            }
        }
       distance++;
      }
     minDistance = Math.min(minDistance, distance-1);
   }
}
